src/Compilers/CSharp/Portable/Binder/LockBinder.cs

```diff
@@ -73,8 +74,18 @@ internal override BoundStatement BindLockStatementParts(BindingDiagnosticBag dia
                     errorCode: ErrorCode.ERR_BadSpecialByRefLock);
             }
 
-            BoundStatement stmt = originalBinder.BindPossibleEmbeddedStatement(_syntax.Statement, diagnostics);
+            var needsFilterDiagnostics = isLockObjectBased && diagnostics.AccumulatesDiagnostics;
+            var bodyDiagnostics = needsFilterDiagnostics ? BindingDiagnosticBag.GetInstance(template: diagnostics) : diagnostics;
+
+            BoundStatement stmt = originalBinder.BindPossibleEmbeddedStatement(_syntax.Statement, bodyDiagnostics);
             Debug.Assert(this.Locals.IsDefaultOrEmpty);
+
+            if (needsFilterDiagnostics)
+            {
+                bodyDiagnostics.CopyFilteredToAndFree(diagnostics,
+                    static code => code is not ErrorCode.WRN_BadYieldInLock);
```

Comment:
While it feels tempting to always report a warning and then filter it out, I would prefer us to not report the warning in the first place. A diagnostic filtering is an exception rather than a usual way of doing things. In this case, I think, we can avoid it. See, for example, how a set of ```LockedOrDisposedVariables``` is exposed and used. I think the fact whether we are in context of a "modern" lock could be accessed/exposed in a similar way, and we could simply not report the warning in the first place.

Reply:
Another thing to consider is to simplify and not attempt to suppress the warning at all. I do not think it is going to lead to any confusion and the problem with the lock object itself is a separate matter.